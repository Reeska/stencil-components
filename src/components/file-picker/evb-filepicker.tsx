import { Component, ComponentInterface, Element, Event, EventEmitter, Listen, Method, Prop } from '@stencil/core';

import { readImage } from '../../utils/file/read-file';

import { PickedFile } from './pickedfile';

@Component({
  tag: 'evb-filepicker',
  styleUrl: 'evb-filepicker.css',
  shadow: true
})
export class EvbFilepicker implements ComponentInterface {

  /** indicates that the user may choose more than one file */
  @Prop() multiple: boolean;

  /**
   * A string that defines the file types the file input should accept. This string is a
   * comma-separated list of unique file type specifiers.
   * To accept images, video and audio, use: accept="audio/*,video/*,image/*"
   * otherwise provide the correct mimetype, eg: image/png for png images etc
   */
  @Prop() accept: string;

  /**
   * Should we show the input type=file?
   */
  @Prop() input = false;

  /**
   * Emits the dataurl for the image
   */
  @Event() pick: EventEmitter<PickedFile>;

  @Element() private host: HTMLElement;

  /**
   * Returns the (hidden) file input element
   */
  get pickerInput(): HTMLInputElement {
    return this.host.shadowRoot.querySelector<HTMLInputElement>('input[type="file"]');
  }

  @Listen('click')
  handleClick(_: MouseEvent) {
    if (!this.input) {
      this.pickerInput.click();
    }
    // otherwise use the input to pick a file
  }

  @Method()
  handleFiles(files: FileList) {
    if (files && files.length) {
      // tslint:disable-next-line: prefer-for-of for loop for FileList
      for (let i = 0; i < files.length; i++) {
        const file = files[i];
        if (this.validateMimes(file, this.accept)) {
          readImage(file)
            .then(dataUrl => {
              this.pick.emit({
                file,
                dataUrl
              });
            });
        }
      }
    }
  }

  /**
   * Validate a file against allowed mimetypes
   */
  validateMimes(file: File, allowedMimeTypes?: string): boolean {

    // check if we want to limit the filepicking...
    if (!allowedMimeTypes) {
      return true;
    }

    if (allowedMimeTypes.indexOf(',') === -1) {
      return this.validateMimeType(file, allowedMimeTypes);
    } else {
      return allowedMimeTypes.split(',').reduce((acc, cur) =>
        this.validateMimeType(file, cur) || acc, false);
    }
  }

  render() {
    return ([
      <input type='file'
        class={this.input ? '' : 'hidden'}
        onChange={event => this.handleFiles((event.target as HTMLInputElement).files)}
        multiple={this.multiple}
        accept={this.accept} />,
      <slot />
    ]);
  }

  private validateMimeType(file: File, mimeType: string): boolean {
    if (file.type === mimeType) {
      return true;
    }

    // handle audio/*, video/*, image/*
    if (['audio/*', 'video/*', 'image/*'].find(item => item === mimeType)) {
      const wildcardRegex = new RegExp(mimeType.replace('/*', '\/.*'));
      return wildcardRegex.test(file.type);
    }

    const regex = new RegExp(mimeType.trim(), 'i');
    return regex.test(file.type);
  }

}
