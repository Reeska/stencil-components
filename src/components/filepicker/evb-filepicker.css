:host {
  display: inline-flex;
  justify-content: flex-start;
}

input[type=file].hidden {
  display: none;
}
