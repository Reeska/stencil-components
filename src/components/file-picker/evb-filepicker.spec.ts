import { EvbFilepicker } from "./evb-filepicker";

describe('evb-filepicker', () => {

  let component: EvbFilepicker;

  beforeEach(() => component = new EvbFilepicker());

  it('should render', () => {
    expect(component.render()).toMatchSnapshot();
  });

  describe('validation', () => {

    describe('validate simple mimetypes', () => {

      it('should validate a mimetype', () => {
        const mimeType = 'image/jpg';
        const mockFile = jest.fn(() => ({ type: mimeType }));
        // @ts-ignore
        const result = component.validateMimes(mockFile(), mimeType);

        expect(result).toBe(true);
      });

      it('should reject a mimetype', () => {
        const mimeType = 'image/jpg';
        const mockFile = jest.fn(() => ({ type: 'image/jepg' }));
        // @ts-ignore
        const result = component.validateMimes(mockFile(), mimeType);

        expect(result).toBe(false);
      });

      it('should accept all when accept=empty', () => {
        const mockFile = jest.fn(() => ({ type: 'image/jpeg' }));
        // @ts-ignore
        const result = component.validateMimes(mockFile());

        expect(result).toBe(true);
      });
    });

    describe('validate wildcard mimetypes', () => {

      it('should validate image/*', () => {
        const mockFile = jest.fn(() => ({
          type: 'image/png'
        }));

        // @ts-ignore
        const result = component.validateMimes(mockFile(), 'image/*');
        expect(result).toBe(true);
      });

      it('should validate audio/*', () => {
        const mockFile = jest.fn(() => ({
          type: 'audio/mp4'
        }));

        // @ts-ignore
        const result = component.validateMimes(mockFile(), 'audio/*');
        expect(result).toBe(true);
      });

      it('should validate video/*', () => {
        const mockFile = jest.fn(() => ({
          type: 'video/mp4'
        }));

        // @ts-ignore
        const result = component.validateMimes(mockFile(), 'video/*');
        expect(result).toBe(true);
      });
    });

    describe('validate compound mimetypes', () => {
      it('should validate mimetypes', () => {
        const mimeTypes = 'mime/type, type/mime';
        const mockFile = jest.fn(() => ({ type: 'type/mime' }));
        // @ts-ignore
        const result = component.validateMimes(mockFile(), mimeTypes);

        expect(result).toBe(true);
      });

      it('should reject mimetypes', () => {
        const mimeTypes = 'mime/types, type/mimes';
        const mockFile = jest.fn(() => ({ type: 'type/mime' }));
        // @ts-ignore
        const result = component.validateMimes(mockFile(), mimeTypes);

        expect(result).toBe(false);
      })
    });

  });

});

