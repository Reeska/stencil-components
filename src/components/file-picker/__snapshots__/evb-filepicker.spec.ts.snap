// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`evb-filepicker should render 1`] = `
Array [
  Object {
    "elm": undefined,
    "ishost": false,
    "vattrs": Object {
      "accept": undefined,
      "class": "hidden",
      "multiple": undefined,
      "onChange": [Function],
      "type": "file",
    },
    "vchildren": null,
    "vkey": undefined,
    "vname": undefined,
    "vtag": "input",
    "vtext": undefined,
  },
  Object {
    "elm": undefined,
    "ishost": false,
    "vattrs": null,
    "vchildren": null,
    "vkey": undefined,
    "vname": undefined,
    "vtag": "slot",
    "vtext": undefined,
  },
]
`;
